### ISSUE TITLE
Separate optimization from GNeuralNet

### ISSUE THREAD

**Issue by thelukester92** [mikegashler/waffles#26]
Right now, all optimization techniques are built into the neural net class and each layer class. Ideally, instead, we would be able to create external optimizers that don't care about layer type aside from the gradient. The idea is that we could do something like this:

    GNeuralNet nn(...);
    GMatrix features(...), labels(...);
    GRMSProp optimizer;
    for(size_t i = 0; i < features.rows(); ++i)
    {
        nn.forwardProp(features[i]);
        nn.backpropagate(labels[i]);
        nn.updateDeltas(features[i]); // this calculates the gradient
        optimizer.trainIncremental(nn.weights(), nn.gradient());
    }

GNeuralNet::weights would be a 1-dimensional vector of all weights in the network (but no copy should be required) and GNeuralNet::gradient would be a 1-dimensional vector of the corresponding derivatives (again with no copies).

This would be a massive, breaking change. It is also not clear how each layer would share the weights given the current architecture. It would be nice if we could get the GNeuralNet to allocate a giant vector and give a pointer to each segment for each layer.

A slightly easier approach would be to make each layer store its weights in a single vector, then have GNeuralNet::weights return a vector of vectors (i.e. vector<GVec>) to optimize. This has the benefit that adding layers doesn't require a full reallocation for the whole network. However, it's still a huge change. Biases would have to be absorbed into the weights matrix, and we would need to guarantee that the weights matrix rows are contiguous (maybe start by allocating a giant vector for the layer and then give that to a GMatrix which splits that vector up into rows).

Thoughts? I just think GNeuralNet is too monolithic, Layers should be easier to add, and the optimizer should be separate from both.

**Comment by StephenAshmore** [comment]
I strongly agree with this. Separating the optimizer logic out from the neural network logic is a very strong feature of a lot of machine learning libraries. It makes our neural networks easier to extend ( want Adam? Just write a new optimizer class, etc ) and easier to maintain. 

Perhaps we should make this a milestone, Waffles 2.0 or some such thing. Then we could even set up #12 for that milestone.

**Comment by thelukester92** [comment]
BTW we have ADAM now as well as RMSProp, but it doesn't work for all layer types yet.

**Comment by thelukester92** [comment]
Agree that we would mark a current version as release before introducing anything for this issue.

**Comment by thelukester92** [comment]
Apparently I had a branch for this about a year ago. I'll drop that branch and make a new one.

**Comment by thelukester92** [comment]
Working on this now!

**Comment by mikegashler** [comment]
Cool. Can this feature somehow subsume the GOptimizer
<http://uaf46365.ddns.uark.edu/waffles/apidoc/html/class_g_classes_1_1_g_optimizer.html>
class?

**Comment by thelukester92** [comment]
Eventually it could, but I think for now I want to make this optimizer separate (too many subclasses to try to reimplement with a different interface).

I have it working for GLayerClassic, now I need to implement a couple of functions for the other layers and make sure it doesn't cause a slowdown.

Is it alright if I break backwards compatibility? GNeuralNet and GNeuralNetLayer will no longer need to store their own deltas, so the interface could be cleaned up by dropping some functions that will be obsolete after this change.

**Comment by mikegashler** [comment]
For science, I'm willing to have to fix my experiments when you break compatibility.

**Comment by thelukester92** [comment]
I have one version that doesn't break the old interface and another (newer) version that will definitely break anything that uses a GNeuralNet. We can either merge them all in together or merge in a little at a time. Any thoughts?

**Comment by thelukester92** [comment]
Here's the list of things affected/changed/removed. Let me know what needs to be fixed and any ideas on how to fix it.

The big problem is GNeuralNet inherits from GSupervisedLearner, so everything expects it to be able to train itself (most problematic is GAutoFilter). One option is to extend the optimizers so that they also work with GSupervisedLearner and we can rip out training from that. Another is to break the connection between GSupervisedLearner and GNeuralNet (but I feel like that will cause another set of problems).

	Affected:
		GLearnerLib::InstantiateNeuralNet (removed most of the functionality)
		GLearnerLib::InstantiateWag (removed most of the NN functionality)
		InstantiateNeuralNet (waffles_sparse)
	
	Definitely broken:
		GLearnerLib::autoTuneNeuralNet
		GAutoFilter
		GNeuralNet_testBinaryClassification (because of GAutoFilter)
		GReservoirNet::test (because of GAutoFilter)
	
	Problems:
		Default is now batch mode, but that slows down several of the tests
	
	Removed:
		GNeuralNet::improvementThresh
		GNeuralNet::windowSize
		GNeuralNet::trainSparse()
		GNeuralNet::trainWithValidation()
		GNeuralNet::internalTrainingGMatrix()
		GNeuralNet::internalValidationData()
		GNeuralNet::copyErrors()
		GNeuralNet::forwardPropSingleOutput()
		GNeuralNet::sumSquaredPredictionError()
		GNeuralNet::autoTune()
		GNeuralNet::backpropagateErrorAlreadySet()
		GNeuralNet::backpropagate() -- replaced with backpropagate(GVec &blame)
		GNeuralNet::backpropagateAndNormalizeErrors()
		GNeuralNet::backpropagateFromLayerAndNormalizeErrors()
		GNeuralNet::backpropagateSingleOutput()
		GNeuralNet::descendGradient()
		GNeuralNet::descendGradientAdam()
		GNeuralNet::descendGradientSingleOutput()
		GNeuralNet::updateDeltas()
		GNeuralNet::applyDeltas()
		GNeuralNet::gradientOfInputs()
		GNeuralNet::gradientOfInputsSingleInput()
		GNeuralNet::trainIncremental()
		GNeuralNet::trainIncrementalBatch()
		GNeuralNet::trainIncrementalWithDropout()
		GNeuralNet::pretrainWithAutoencoders()
		GNeuralNet::trainInner()
		GNeuralNet::validationSquaredError()
		GNeuralNet_testConvolutionalLayerMath()
		GNeuralNet_testInputGradient()
		GNeuralNet_testConvenience()
		GBackPropThroughTime
		GNeuralNetLayer::computeError()
		GNeuralNetLayer::updateDeltas() -- replaced with updateDeltas(GVec &deltas)
		GNeuralNetLayer::applyDeltas() -- replaced with applyDeltas(GVec &deltas)
		GLayerClassic::updateDeltasAdam()
		GLayerClassic::applyDeltasAdam()
		GLayerClassic::computeErrorSingleOutput()
		GLayerClassic::updateWeightsSingleNeuron()
		GAttributeSelector::trainInner()

**Comment by mikegashler** [comment]
[inline]

On 10/27/2016 03:58 PM, Luke Godfrey wrote:
>
> Here's the list of things affected/changed/removed. Let me know what
> needs to be fixed and any ideas on how to fix it.
>
> The big problem is GNeuralNet inherits from GSupervisedLearner, so
> everything expects it to be able to train itself (most problematic is
> GAutoFilter).
>
Can't we just pick a default optimizer and make a train method that uses
it? I am happy to nuke GAutoFilter for neural nets. (It was too slow to
be useful anyway.)

> One option is to extend the optimizers so that they also work with
> GSupervisedLearner and we can rip out training from that. Another is
> to break the connection between GSupervisedLearner and GNeuralNet (but
> I feel like that will cause another set of problems).
>
> |Affected: GLearnerLib::InstantiateNeuralNet (removed most of the
> functionality) GLearnerLib::InstantiateWag (removed most of the NN
> functionality) InstantiateNeuralNet (waffles_sparse) Definitely
> broken: GLearnerLib::autoTuneNeuralNet|
Good riddance

> |GAutoFilter|
|I think this will fix itself when we make "train" work again, right?

|
> |GNeuralNet_testBinaryClassification (because of GAutoFilter)
> GReservoirNet::test (because of GAutoFilter)|
I don't care.

||
> |Problems: Default is now batch mode, but that slows down several of
> the tests|
Stochastic should definitely be the default. Is that problematic for any
reason?

> |Removed: GNeuralNet::improvementThresh GNeuralNet::windowSize
> GNeuralNet::trainSparse()|
|Good riddance.

|||||
> |GNeuralNet::trainWithValidation()|
I don't care about preserving this interface, but the new optimizer
class definitely needs to implement some way to detect convergence using
a validation set. Hopefully, this can be done in a general way that
works for all optimization methods.

> |GNeuralNet::internalTrainingGMatrix()
> GNeuralNet::internalValidationData() GNeuralNet::copyErrors()|
Good riddance.

> |GNeuralNet::forwardPropSingleOutput()|
Can some new form of this be salvadged? It is needed by one of my
collaborative filters.

> |GNeuralNet::sumSquaredPredictionError() GNeuralNet::autoTune()
> GNeuralNet::backpropagateErrorAlreadySet() GNeuralNet::backpropagate()
> -- replaced with backpropagate(GVec &blame)
> GNeuralNet::backpropagateAndNormalizeErrors()
> GNeuralNet::backpropagateFromLayerAndNormalizeErrors()
> GNeuralNet::backpropagateSingleOutput() GNeuralNet::descendGradient()
> GNeuralNet::descendGradientAdam()
> GNeuralNet::descendGradientSingleOutput() GNeuralNet::updateDeltas()
> GNeuralNet::applyDeltas() GNeuralNet::gradientOfInputs()
> GNeuralNet::gradientOfInputsSingleInput()|
I think the new approach provides equivalents for all of these, and that
is good enough.

> |GNeuralNet::trainIncremental()|
I'd like to save this interface if possible.

> |GNeuralNet::trainIncrementalBatch()|
|Never heard of that. Isn't that an oxymoron?

|
> |GNeuralNet::trainIncrementalWithDropout()
> GNeuralNet::pretrainWithAutoencoders() GNeuralNet::trainInner()
> GNeuralNet::validationSquaredError()
> GNeuralNet_testConvolutionalLayerMath() GNeuralNet_testInputGradient()
> GNeuralNet_testConvenience() GBackPropThroughTime
> GNeuralNetLayer::computeError() GNeuralNetLayer::updateDeltas() --
> replaced with updateDeltas(GVec &deltas)
> GNeuralNetLayer::applyDeltas() -- replaced with applyDeltas(GVec
> &deltas) GLayerClassic::updateDeltasAdam()
> GLayerClassic::applyDeltasAdam()|
I'm okay with losing this stuff.

> |GLayerClassic::computeErrorSingleOutput()
> GLayerClassic::updateWeightsSingleNeuron()|
These are part of the collaborative filter thing.

> |GAttributeSelector::trainInner()|
This class should be redesigned, but I think we should definitely fix
it. It's very useful.

**Comment by thelukester92** [comment]
Sounds good. I have implemented optimizeWithValidation, and the other changes should be reasonable. I'll work on it.

**Comment by thelukester92** [comment]
Good news: the new optimizer apparently works better than the old one. When I run the test now:

    The measured accuracy (0.789) is much better than expected (0.75). Please increase the expected accuracy value so that any future regressions will be caught.

**Comment by thelukester92** [comment]
So you want to keep GLayerClassic::computeErrorSingleOutput() and GLayerClassic::updateWeightsSingleNeuron()? Can you send me the code that uses it?

I'm not sure how to reimplement the GNeuralNet methods that use them (i.e. descendGradientSingleOutput) as part of the separate optimizer.

**Comment by thelukester92** [comment]
Training is no longer slow, accuracy is passing all tests, and the GSupervisedLearner interface is working now that we have a default SGD optimizer in GNeuralNet. I think this change is about finished. It's pushed in a branch right now if you want to test it out.

**Comment by thelukester92** [comment]
I updated the web docs (neuralnet.html) with instructions on how to optimize the new way. Replace GSGDOptimizer with GRMSPropOptimizer if desired.

**Comment by thelukester92** [comment]
Will implement #31 before merging. For the forwardPropOneOutput and friends, we could make a separate GNeuralNetSingleOutputFunction (target function) to optimize instead of GNeuralNetFunction.

**Comment by thelukester92** [comment]
Need to implement slack.

**Comment by thelukester92** [comment]
Done and ready to merge. Any objections?

**Comment by thelukester92** [comment]
Merged.
